lib/syskit/log/datastore/normalize.rb

```diff
@@ -108,20 +110,65 @@ def create_block_stream
                     Pocolog::BlockStream.new(@wio.dup)
                 end
 
+                def update_raw_payload_logical_time(raw_payload, logical_time)
+                    # Logical time are bytes from 8..15
+                    raw_payload[8..11] = [logical_time.tv_sec].pack("V")
+                    raw_payload[12..15] = [logical_time.tv_usec].pack("V")
+                    raw_payload
+                end
+
                 def add_data_block(rt_time, lg_time, raw_data, raw_payload)
                     @stream_size += 1
 
                     write raw_data[0, 2]
                     write ZERO_BYTE
                     write raw_data[4..-1]
+
+                    logical_time = extract_logical_time(raw_payload)
+                    if logical_time
+                        lg_time = logical_time.microseconds
+                        raw_payload = update_raw_payload_logical_time(
+                            raw_payload, logical_time
+                        )
+                    end
                     write raw_payload
+
                     @interval_rt[0] ||= rt_time
                     @interval_rt[1] = rt_time
                     @interval_lg[0] ||= lg_time
                     @interval_lg[1] = lg_time
                     @last_data_block_time = [rt_time, lg_time]
                 end
 
+                def resolve_logical_time_field(stream_block)
+                    rock_timestamp_field = stream_block.metadata["rock_timestamp_field"]
+                    return rock_timestamp_field if rock_timestamp_field
+
+                    type = stream_block.type
+                    return unless type < Typelib::CompoundType
+
+                    metadata = type.field_metadata
+                    type.each_field do |field|
+                        next unless metadata[field].include?("role")
```

Comment:
`.get()` returns an empty array if a metadata key is not included. This check is not needed.